File: web/controllers/api/news_controller.ex
defmodule Scanner.API.NewsController do
  use Scanner.Web, :controller
  use Scanner.JSONHelper

  alias Scanner.News

  defdelegate fetch_all_news(conn, _options), to: Scanner.NewsController
  plug :fetch_all_news when action == :index

  def index conn, params do
    %{news: news, page: page, finished: finished} = conn.assigns
    json_success conn, %{data: news, page: page, finished: finished}
  end
end

File: app/controllers/api/news_controller.rb
class Api::NewsController < Api::ApplicationController
  def index
    @page = (params[:page] || 1).to_i
    @per = (params[:per] || 10).to_i

    news_klass =
      if params[:category]
        News.where(category: params[:category])
      else
        News
      end

    @finished = news_klass.count <= @page * 10
    @news = news_klass.page(@page).per(@per)
  end
end
